core-client/src/transports/duplex.rs

```diff
@@ -0,0 +1,126 @@
+//! Duplex transport
+
+use failure::format_err;
+use futures::prelude::*;
+use futures::sync::{mpsc, oneshot};
+use jsonrpc_core::Id;
+use log::debug;
+use serde_json::Value;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+
+use super::RequestBuilder;
+use crate::{RpcChannel, RpcError, RpcMessage};
+
+/// The Duplex handles sending and receiving asynchronous
+/// messages through an underlying transport.
+pub struct Duplex<TSink, TStream> {
+	request_builder: RequestBuilder,
+	queue: HashMap<Id, oneshot::Sender<Result<Value, RpcError>>>,
+	sink: TSink,
+	stream: TStream,
+	channel: Option<mpsc::Receiver<RpcMessage>>,
+	outgoing: VecDeque<String>,
+}
+
+impl<TSink, TStream> Duplex<TSink, TStream> {
+	/// Creates a new `Duplex`.
+	fn new(sink: TSink, stream: TStream, channel: mpsc::Receiver<RpcMessage>) -> Self {
+		Duplex {
+			request_builder: RequestBuilder::new(),
+			queue: HashMap::new(),
+			sink,
+			stream,
+			channel: Some(channel),
+			outgoing: VecDeque::new(),
+		}
+	}
+
+	/// Creates a new `Duplex`, along with a channel to communicate
+	pub fn with_channel(sink: TSink, stream: TStream) -> (Self, RpcChannel) {
+		let (sender, receiver) = mpsc::channel(0);
+		let client = Duplex::new(sink, stream, receiver);
+		(client, sender.into())
+	}
+}
+
+impl<TSink, TStream> Future for Duplex<TSink, TStream>
+where
+	TSink: Sink<SinkItem = String, SinkError = RpcError>,
+	TStream: Stream<Item = String, Error = RpcError>,
+{
+	type Item = ();
+	type Error = RpcError;
+
+	fn poll(&mut self) -> Result<Async<Self::Item>, Self::Error> {
+		// Handle requests from the client.
+		loop {
+			if self.channel.is_none() {
+				break;
+			}
+			let msg = match self.channel.as_mut().expect("channel is some; qed").poll() {
+				Ok(Async::Ready(Some(msg))) => msg,
+				Ok(Async::Ready(None)) => {
+					// When the channel is dropped we still need to finish
+					// outstanding requests.
+					self.channel.take();
+					break;
+				}
+				Ok(Async::NotReady) => break,
+				Err(()) => continue,
+			};
+			let (id, request_str) = self.request_builder.single_request(&msg);
+			self.queue.insert(id, msg.sender);
+			self.outgoing.push_back(request_str);
+		}
+		// Handle outgoing rpc requests.
+		loop {
+			match self.outgoing.pop_front() {
+				Some(request) => match self.sink.start_send(request)? {
```

Comment:
Wouldn't it be easier to use `send_all` instead?

Reply:
Task in https://github.com/paritytech/jsonrpc/issues/432